src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/Policies/CamelCasingPolicy.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization.Policies
+{
+    internal static class CamelCasingPolicy
+    {
+        public static string FromJson(string value)
+        {
+            if (value.Length == 0 || char.IsUpper(value[0]))
+                return value;
+
+            if (value.Length == 1)
+                return value.ToUpperInvariant();
+
+            return char.ToUpperInvariant(value[0]) + value.Substring(1);
```

Comment:
Can we use span/slicing to avoid unnecessary substring allocations?

Reply:
Not in this case as I want to return a string

Reply:
You are able to return a string, but without allocation of two additional strings which are concatenated to ptoduce the result. Use the `string.Create<TState>(Int32, TState, SpanAction<Char, TState>)`method instead and pass `value` as a state object. This approach allows you to remove the second condition from the method.

Reply:
Current:
Allocations: 88 bytes
Runtime: 39 ns

You could use the SpanAction, which is faster (though it allocates more):
```C#
string result = string.Create(value.Length, first, (Span<char> chars, char ch) =>
{
    chars[0] = char.ToUpperInvariant(ch);
    value.AsSpan(1).CopyTo(chars.Slice(1));
});
```
Allocations: 120 bytes
Runtime: 31 ns

OR, if we really care, use unsafe code (which is much faster with no allocations):
```C#
ReadOnlySpan<char> span = value.AsSpan();
Span<char> writableSpan = MemoryMarshal.CreateSpan(ref MemoryMarshal.GetReference(span), span.Length);
writableSpan[0] = char.ToUpperInvariant(first);
return new string(writableSpan);
```
Allocations: 0 bytes
Runtime: 4 ns

If we want to only use safe code, we could leverage stackalloc span for small lengths to avoid the allocation:

```C#
public static string FromJsonNew(string value)
{
    if (value.Length == 0)
        return value;

    char first = value[0];
    if (char.IsUpper(first))
        return value;

    if (value.Length == 1)
        return value.ToUpperInvariant();

    if (value.Length < 256)
    {
        Span<char> span = stackalloc char[value.Length];
        span[0] = char.ToUpperInvariant(first);
        value.AsSpan(1).CopyTo(span.Slice(1));
        return new string(span);
    }
    else
    {
        return char.ToUpperInvariant(first) + value.Substring(1);
    }
}
```
Allocations: 32 bytes
Runtime: 33 ns

Reply:
When targeting .NET Core 3, it can be:
```C#
char c = char.ToUpperInvariant(first);
return string.Concat(MemoryMarshal.CreateReadOnlySpan(ref c, 1), value.AsSpan(1));
```

Reply:
> You could use the SpanAction, which is faster (though it allocates more):

It doesn't need to allocate beyond the returned string (other than the one-time allocation for the compiler-cached delegate):
```C#
return string.Create(value.Length, (ch, value), (chars, args) =>
{
    chars[0] = char.ToUpperInvariant(args.ch);
    args.value.AsSpan(1).CopyTo(chars.Slice(1));
});
```
You just need to make sure not to capture anything into the closure.

Reply:
`string.Concat` isn't the best for perf here and it ends up using unsafe code, so if we are using `MemoryMarshal`, maybe we shouldn't use `string.Concat` and just use the span-based string ctor.

I am surprised the string ctor doesn't allocate at all.

Reply:
> string.Concat isn't the best for perf here and it ends up using unsafe code

How so?  It should be the fastest of all the options.  It's also not doing anything unsafe; the only thing bordering on that is the MemoryMarshal.CreateReadOnlySpan, and the only thing there that needs to be accurate is the 1 passed in as the length.

> I am surprised the unsafe sample which uses the string ctor doesn't allocate at all.

It allocates the result string like all the others.  What other allocations were you expecting?

Reply:
> Please don't do this: you're mutating the original string

Lol, you are right :)

Yes, we shouldn't do this. 

> It allocates the result string like all the others. What other allocations were you expecting?

The benchmark showed 0 allocations, because we are mutating the original string and the ctor ends up not allocating.

> How so? It should be the fastest of all the options.

Let me try to call the method in a loop, but the SpanAction-based impl took 19 ns, while string.Concat took 30 ns.

Reply:
> but the SpanAction-based impl took 19 ns, while string.Concat took 30 ns.

Sounds like we should improve string.Concat then.  It's probably suffering from doing a Span.CopyTo on the 1-char first span.

Reply:
> The benchmark showed 0 allocations, because we are mutating the original string and the ctor ends up not allocating.

The `return new string(writableSpan);` would still end up allocating.  You're saying it somehow wasn't?

Reply:
``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview-010174
  [Host]     : .NET Core 3.0.0-preview-27324-5 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview-27324-5 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT


```
|       Method |     Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------- |---------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|   SpanAction | 20.88 us | 0.2818 us | 0.2353 us |  0.67 |    0.02 |      9.3689 |           - |           - |            38.28 KB |
| StringConcat | 30.78 us | 0.6499 us | 0.7981 us |  1.00 |    0.00 |      9.3384 |           - |           - |            38.28 KB |

Reply:
> Yes.

What is the implementation of UnsafeMutateString?  It's doing `return new string(writeableSpan)` and it's not allocating?  I don't believe it.

Reply:
> I don't believe it.

> each subsequent call to the method returns the same (unchanged) string so it doesn't allocate
> We never end up calling the string ctor.

You were right not to :)

The results make sense now.

Reply:
I used the AsSpan \ CopyTo approach